### PR TITLE
fix: Added _libexec_prefix variable to NM module-setup.sh for distros w/o libexec

### LINKED ISSUE #1161: network-manager has files missing on Arch Linux (+ other distros without /usr/libexec)
**Describe the bug**
Including the network-manager dracut module on Arch Linux doesn't work fully; as Arch doesn't use libexec, both of /usr/lib/nm-dhcp-helper and /usr/lib/nm-initrd-generator are left out of the initrd.

**Distribution used**
Arch Linux

**Dracut version**
dracut-053

**Init system**
systemd

**To Reproduce**
Add `add_dracutmodules+=" network-manager"` to dracut.conf/dracut.conf.d on an Arch system (or on any system missing /usr/libexec) and generate an initramfs.

**Expected behavior**
Including the network-manager dracut module and then generating the initramfs works just fine, with no missing files.

**Additional context**
This is easily worked around by a quick symlink from libexec to lib.

### PR DESCRIPTION
**Pull request by rosieeyes**
Added _libexec_prefix variable to NM module-setup.sh for distros (e.g. Arch Linux) that don't use /usr/libexec

This pull request changes...

## Changes

When added to the initramfs, the network-manager module will miss the nm-dhcp-helper and nm-initrd-generator files on distros such as Arch Linux that install files which would otherwise go into /usr/libexec, to /usr/lib.

This changes module-setup.sh to use /usr/lib if /usr/libexec doesn't exist.

This assumes that if the two NM files can't be found under /usr/libexec, that they'll be under /usr/lib instead, which I think is fair - it works for Arch Linux, at least.

I'm definitely welcome to hear better variable names than "_libexec_prefix".

## Checklist
- [x] I have tested it locally
- [ ] I have reviewed and updated any documentation if relevant
- [ ] I am providing new code and test(s) for it

Fixes #1161

### CODE REVIEW

**Review by johannbg — CHANGES_REQUESTED** [review]
You need to introduce this change as an global libexecdir variable not per module fix and then each module that uses /usr/libexec be fixed to use that global libexecdir variable

At that point each distro can just configure $libexecdir=<it's distro path> in it's dracut distribution configuration file

Basically the scope of this change is dracut.sh, 35network-manager, 40network, 80cms,91crypt-gpg, 35network-wicked,50plymouth modules along with a downstream patch to Arch default dracut.conf

**Comment by rosieeyes** [comment]
Thanks for the input! I definitely agree that a global config option is best, but there are some modules that already handle both /usr/lib and /usr/libexec where necessary. Is it worth changing these instances when they're already working fine? Context below...

* [35network-manager](https://github.com/dracutdevs/dracut/blob/cc914eaa8b292e846a3a794abe9ef40db6eab7b5/modules.d/35network-manager/nm-lib.sh#L7) has another instance in which I'll change as discussed
* [35network-wicked](https://github.com/dracutdevs/dracut/blob/cc914eaa8b292e846a3a794abe9ef40db6eab7b5/modules.d/35network-wicked/module-setup.sh#L43) installs from both /usr/lib and /usr/libexec
* [40network](https://github.com/dracutdevs/dracut/blob/cc914eaa8b292e846a3a794abe9ef40db6eab7b5/modules.d/40network/module-setup.sh#L22) needs changing much like 35network-manager
* [50plymouth](https://github.com/dracutdevs/dracut/blob/cc914eaa8b292e846a3a794abe9ef40db6eab7b5/modules.d/50plymouth/module-setup.sh#L4) looks in both of /usr/lib and /usr/libexec already
* 80cms; I couldn't find any references to /usr/libexec there
* [91crypt-gpg](https://github.com/dracutdevs/dracut/blob/cc914eaa8b292e846a3a794abe9ef40db6eab7b5/modules.d/91crypt-gpg/module-setup.sh#L48) just uses /usr/libexec, so I'll change this

I'm also not keen on testing each module where I modify $libexecdir; I hope that just testing 35network-manager will suffice.

**Comment by johannbg** [comment]
> Thanks for the input! I definitely agree that a global config option is best, but there are some modules that already handle both /usr/lib and /usr/libexec where necessary. Is it worth changing these instances when they're already working fine?

When introducing global variable we should replace all the existing instances to use that global variable instead.  

>     * 80cms; I couldn't find any references to /usr/libexec there

It's in cmsifup.sh
```
if [ -x /usr/libexec/nm-initrd-generator ]; then
...
```
Note when you add the global variable in dracut.sh which will be something along the lines of ... 

```
[[ -d $dracutsysrootdir$libexecdir ]] \
    || libexecdir=$(pkg-config libexecdir --variable=libexecdir 2> /dev/null)
[[ -d $dracutsysrootdir$libexecdir ]] || libexecdir=/usr/libexec
```

Dont forget to add your global variable to the export line
```
export initdir dracutbasedir \
....
tmpfilesdir <your global variable>
```

You also need to fix your commit messages to pass the commisery check which will be something along the lines of ... 

```
feat(dracut.sh): introducing <your global variable> 
....
```
for the dracut .sh file and for modules we dont include the module numbers since those are subjected to changes so it would look like something like... 

```
fix(network-manager): use <your global variable > instead of path
...
```
and 
```
fix(network-wicked): use <your global variable > instead of path
...
```
etc.

**Comment by rosieeyes** [comment]
That's mostly everything taken care of, though I'm struggling to find where to add the necessary downstream changes to change libexecdir. I think it's best to add it as a part of ./configure, though I don't see where to make changes to facilitate that... it seems as though there's already been some work in the Makefile and configure files, but I don't see where to go next with it. I'll keep working on it, but feel free to direct me if you wish. :)

I also squashed everything together in order to fix the Commisery issue, but it seems that doesn't work. Does each file change need to be its own commit?

**Comment by johannbg** [comment]
> That's mostly everything taken care of, though I'm struggling to find where to add the necessary downstream changes to change libexecdir from its default. I think it's best to add it as a part of ./configure, though I don't see where to make changes to facilitate that... it seems as though there's already been some work in the Makefile and configure files, but I don't see where to go next with it. I'll keep working on it, but feel free to direct me if you wish. :)

I simply suggest you wait a bit with that one until this is merged since we might end up actually using the systemd file-hierarchy which Arch actually defaults to so you might end up having to do nothing downstream. 

> 
> I also squashed everything together + amended the commit message in order to fix the Commisery issue, but it seems that doesn't work. Does each file change need to be its own commit?

Either that or better yet in two commits one for the feature ```feat: introducing libexecdir global variable``` and another squashed commits for all the fixes  under ```fix: use libexecdir global variable``` 

I suggest you read how to write a good commit message which can be found here [1] before fixing the commit message since you seem to be somewhat rushing it. 

This wont be merged until it has passed the second reviewer and he's not available until sometime tomorrow so there's no rush so just take your time.  

https://chris.beams.io/posts/git-commit/

**Comment by rosieeyes** [comment]
> > That's mostly everything taken care of, though I'm struggling to find where to add the necessary downstream changes to change libexecdir from its default. I think it's best to add it as a part of ./configure, though I don't see where to make changes to facilitate that... it seems as though there's already been some work in the Makefile and configure files, but I don't see where to go next with it. I'll keep working on it, but feel free to direct me if you wish. :)
> 
> I simply suggest you wait a bit with that one until this is merged since we might end up actually using the systemd file-hierarchy which Arch actually defaults to so you might end up having to do nothing downstream.

Alrighty, is there a PR or issue for that? It'd be useful to keep tabs on and see how that progresses.

> Either that or better yet in two commits one for the feature `feat: introducing libexecdir global variable` and another squashed commits for all the fixes under `fix: use libexecdir global variable`
> 
> I suggest you read how to write a good commit message which can be found here [1] before fixing the commit message since you seem to be somewhat rushing it.
> 
> This wont be merged until it has passed the second reviewer and he's not available until sometime tomorrow so there's no rush so just take your time.
> 
> https://chris.beams.io/posts/git-commit/

Hah, you could tell I was rushing... I just thought this would be a quick fix I could push out on a night and have done with it, but here we are. I've never had to have my commit messages pass linting as much my code does; thanks for the reference.

**Comment by johannbg** [comment]
> > I simply suggest you wait a bit with that one until this is merged since we might end up actually using the systemd file-hierarchy which Arch actually defaults to so you might end up having to do nothing downstream.
> 
> Alrighty, is there a PR or issue for that? It'd be useful to keep tabs on and see how that progresses.

Well it's in your PR the change would happen and it would be made by you. 

It's something we need to discuss among ourselves how it affects Fedora/OpenSuse if we do, so adjustment there can be made accordingly at the same time.

> Hah, you could tell I was rushing... 

Well I have been uhum guilty of doing it myself couple of times in the past so I know all the signs :smile:

File: modules.d/35network-manager/nm-lib.sh
Comment (haraldh) [inline modules.d/35network-manager/nm-lib.sh:7]:
```suggestion
    "$libexecdir"/nm-initrd-generator -- $(getcmdline)
```

File: modules.d/35network-manager/module-setup.sh
Comment (haraldh) [inline modules.d/35network-manager/module-setup.sh:49]:
```suggestion
        inst "$libexecdir"/nm-dhcp-helper
```

File: modules.d/35network-manager/module-setup.sh
Comment (haraldh) [inline modules.d/35network-manager/module-setup.sh:34]:
```suggestion
    inst "$libexecdir"/nm-initrd-generator
```

File: modules.d/80cms/cmsifup.sh
Comment (haraldh) [inline modules.d/80cms/cmsifup.sh:50]: there is no `$libexecdir` in the initramfs root... yet

File: modules.d/91crypt-gpg/module-setup.sh
Comment (haraldh) [inline modules.d/91crypt-gpg/module-setup.sh:32]:
```suggestion
        inst_multiple "$libexecdir"/scdaemon
```

File: modules.d/91crypt-gpg/module-setup.sh
Comment (haraldh) [inline modules.d/91crypt-gpg/module-setup.sh:48]:
```suggestion
        && require_binaries "$libexecdir"/scdaemon \
```

File: modules.d/35network-manager/nm-lib.sh
Comment (haraldh) [reply modules.d/35network-manager/nm-lib.sh:7]: .. but... there is no $libexecdir in the initramfs root... yet

**Review by haraldh — CHANGES_REQUESTED** [review]
I would prefer to install all `$libexedir` to a standard location like `/usr/bin`, so that we don't have to distinguish distro specific stuff in the initramfs runtime...

E.g. install `nm-initrd-generator` to `/usr/bin` and call it from there, instead of introducing a `libexecdir` global variable in the initramfs to be read at runtime.

**Comment by johannbg** [comment]
> I would prefer to install all `$libexedir` to a standard location like `/usr/bin`, so that we don't have to distinguish distro specific stuff in the initramfs runtime...

Given the needless fragmentation on the core/baseOS layer between distribution I dont think we want to end up with a situation where paths in the initramfs differ from the vendor/distribution so we have to provide means to accommodate for it, which is best done using global variables as much as possible in modules. 

> 
> E.g. install `nm-initrd-generator` to `/usr/bin` and call it from there, instead of introducing a `libexecdir` global variable in the initramfs to be read at runtime.

The purpose of `/usr/libexec` was to host internal binaries that are not intended to be executed directly by users or shell scripts so they dont belong in /usr/bin 

FHS and SD-FHS refer to `/usr/lib` should be used for that purpose ( FHS states that `/usr/libexec` can also "optionally" be used for that purpose  ) so we should be defaulting to use `/usr/lib` for this not `/usr/bin` to be following the FHS's.

**Comment by Conan-Kudo** [comment]
Both openSUSE and Debian have since migrated to `/usr/libexec`, so I'm a little surprised Arch hasn't made that change too.

File: dracut.sh
Comment (haraldh) [inline dracut.sh:1742]: There is no such thing as a `libexecdir` pkg-config...

```suggestion
if ! [[ -d $dracutsysrootdir$libexecdir ]]; then
    if [[ -d $dracutsysrootdir/usr/libexec ]]; then
        libexecdir=/usr/libexec
    else
        libexecdir=/usr/lib
    fi
fi    
```

**Review by haraldh — CHANGES_REQUESTED** [review]

**Comment by haraldh** [comment]
Fixed with https://github.com/dracutdevs/dracut/pull/1480 and https://github.com/dracutdevs/dracut/pull/1473

**Pull request closed without merging**
